src/nm-checkpoint.c

```diff
@@ -433,6 +463,7 @@ device_checkpoint_create (NMCheckpoint *checkpoint, NMDevice *device)
 	dev_checkpoint->device = g_object_ref (device);
 	dev_checkpoint->original_dev_path = g_strdup (path);
```

Comment:
there is actually little point in preserving the `original_dev_path`.

Each path MUST be generated uniquely, and we should never export an `NMDBusObject` (like `NMDevice`) more than once. The D-Bus path ("exported path") is only useful as an ID when communicating with the user (== the D-Bus API). After the NMDevice gets unexported (the first and only time), the path becomes meaningless.

In particular:
```
 next_dev:
          g_variant_builder_add (&builder, "{su}", dev_checkpoint->original_dev_path, result);
```
is wrong.

Either the device is still exported (in which case you can just use `nm_exported_object_get_path()`) -- or it is not and then it is a bug to tell the user about a device that no longer exists.

Reply:
But how do you report that a removed device has not been restored in the rollback?